// protractor-flake/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const LOG_LEVELS = {
    debug: 0,
    protractor: 1,
    info: 2,
    silent: 3
};
class Logger {
    constructor(color) {
        this.color = color;
    }
    log(levelName, message, useColor = true) {
        let logLevel = process.env.PROTRACTOR_FLAKE_LOG_LEVEL;
        let currentLevel = LOG_LEVELS[logLevel] || LOG_LEVELS.protractor;
        let incomingLevel = LOG_LEVELS[levelName];
        if (incomingLevel >= currentLevel) {
            message = this.colorize(message, useColor);
            process.stdout.write(message);
        }
    }
    protractor(message) {
        this.log('protractor', message, false);
    }
    colorize(message, useColor) {
        if (useColor && chalk_1.default.supportsColor && this.color) {
            return chalk_1.default.keyword(this.color)(message);
        }
        else {
            return message;
        }
    }
}
exports.default = Logger;
