// webdriver-image-comparison/build/methods/images.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = require("chalk");
var fs_extra_1 = require("fs-extra");
var path_1 = require("path");
var compareImages_1 = require("../resemble/compareImages");
var utils_1 = require("../helpers/utils");
var constants_1 = require("../helpers/constants");
var rectangles_1 = require("./rectangles");
var _a = require('canvas'), createCanvas = _a.createCanvas, loadImage = _a.loadImage;
function checkBaselineImageExists(actualFilePath, baselineFilePath, autoSaveBaseline) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2, new Promise(function (resolve, reject) {
                    fs_extra_1.access(baselineFilePath, function (error) {
                        if (error) {
                            if (autoSaveBaseline) {
                                try {
                                    fs_extra_1.copySync(actualFilePath, baselineFilePath);
                                    console.log(chalk_1.yellow("\n#####################################################################################\n INFO: \n Autosaved the image to \n " + baselineFilePath + "\n#####################################################################################\n"));
                                }
                                catch (error) {
                                    reject(chalk_1.red("\n#####################################################################################\n Image could not be copied. The following error was thrown: \n " + error + "\n#####################################################################################\n"));
                                }
                            }
                            else {
                                reject(chalk_1.red("\n#####################################################################################\n Baseline image not found, save the actual image manually to the baseline.\n The image can be found here:\n " + actualFilePath + "\n If you want the module to auto save a non existing image to the baseline you\n can provide 'autoSaveBaseline: true' to the options.\n#####################################################################################\n"));
                            }
                        }
                        resolve();
                    });
                })];
        });
    });
}
exports.checkBaselineImageExists = checkBaselineImageExists;
function executeImageCompare(executor, options, isViewPortScreenshot) {
    if (isViewPortScreenshot === void 0) { isViewPortScreenshot = false; }
    return __awaiter(this, void 0, void 0, function () {
        var debug, devicePixelRatio, fileName, isAndroidNativeWebScreenshot, platformName, _a, actualFolder, autoSaveBaseline, baselineFolder, browserName, deviceName, diffFolder, isMobile, savePerInstance, diffFilePath, imageCompareOptions, createFolderOptions, actualFolderPath, baselineFolderPath, actualFilePath, baselineFilePath, resembleIgnoreDefaults, ignore, blockOut, statusAddressToolBarOptions, ignoredBoxes, _b, _c, compareOptions, data, misMatchPercentage, isDifference, isDifferenceMessage, debugMessage, diffFolderPath, _d;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    debug = options.debug, devicePixelRatio = options.devicePixelRatio, fileName = options.fileName, isAndroidNativeWebScreenshot = options.isAndroidNativeWebScreenshot, platformName = options.platformName;
                    _a = options.folderOptions, actualFolder = _a.actualFolder, autoSaveBaseline = _a.autoSaveBaseline, baselineFolder = _a.baselineFolder, browserName = _a.browserName, deviceName = _a.deviceName, diffFolder = _a.diffFolder, isMobile = _a.isMobile, savePerInstance = _a.savePerInstance;
                    imageCompareOptions = __assign(__assign({}, options.compareOptions.wic), options.compareOptions.method);
                    createFolderOptions = { browserName: browserName, deviceName: deviceName, isMobile: isMobile, savePerInstance: savePerInstance };
                    actualFolderPath = utils_1.getAndCreatePath(actualFolder, createFolderOptions);
                    baselineFolderPath = utils_1.getAndCreatePath(baselineFolder, createFolderOptions);
                    actualFilePath = path_1.join(actualFolderPath, fileName);
                    baselineFilePath = path_1.join(baselineFolderPath, fileName);
                    return [4, checkBaselineImageExists(actualFilePath, baselineFilePath, autoSaveBaseline)];
                case 1:
                    _e.sent();
                    resembleIgnoreDefaults = ['alpha', 'antialiasing', 'colors', 'less', 'nothing'];
                    ignore = resembleIgnoreDefaults.filter(function (option) {
                        return Object.keys(imageCompareOptions).find(function (key) {
                            return key.toLowerCase().includes(option) && imageCompareOptions[key];
                        });
                    });
                    blockOut = 'blockOut' in imageCompareOptions ? imageCompareOptions.blockOut : [];
                    statusAddressToolBarOptions = {
                        isMobile: isMobile,
                        isViewPortScreenshot: isViewPortScreenshot,
                        platformName: platformName,
                        isAndroidNativeWebScreenshot: isAndroidNativeWebScreenshot,
                        blockOutStatusBar: imageCompareOptions.blockOutStatusBar,
                        blockOutToolBar: imageCompareOptions.blockOutToolBar,
                    };
                    _c = (_b = blockOut).concat;
                    return [4, rectangles_1.determineStatusAddressToolBarRectangles(executor, statusAddressToolBarOptions)];
                case 2:
                    ignoredBoxes = _c.apply(_b, [_e.sent()]).map(function (rectangles) {
                        return utils_1.calculateDprData({
                            bottom: rectangles.y + rectangles.height,
                            right: rectangles.x + rectangles.width,
                            left: rectangles.x,
                            top: rectangles.y,
                        }, devicePixelRatio);
                    });
                    compareOptions = __assign({ ignore: ignore }, (ignoredBoxes.length > 0 ? { output: { ignoredBoxes: ignoredBoxes } } : {}));
                    return [4, compareImages_1.default(fs_extra_1.readFileSync(baselineFilePath), fs_extra_1.readFileSync(actualFilePath), compareOptions)];
                case 3:
                    data = _e.sent();
                    misMatchPercentage = imageCompareOptions.rawMisMatchPercentage
                        ? data.rawMisMatchPercentage
                        : Number(data.rawMisMatchPercentage.toFixed(2));
                    if (!(misMatchPercentage > imageCompareOptions.saveAboveTolerance || debug)) return [3, 6];
                    isDifference = misMatchPercentage > imageCompareOptions.saveAboveTolerance;
                    isDifferenceMessage = 'WARNING:\n There was a difference. Saved the difference to';
                    debugMessage = 'INFO:\n Debug mode is enabled. Saved the debug file to:';
                    diffFolderPath = utils_1.getAndCreatePath(diffFolder, createFolderOptions);
                    diffFilePath = path_1.join(diffFolderPath, fileName);
                    _d = saveBase64Image;
                    return [4, addBlockOuts(Buffer.from(data.getBuffer()).toString('base64'), ignoredBoxes)];
                case 4: return [4, _d.apply(void 0, [_e.sent(),
                        diffFilePath])];
                case 5:
                    _e.sent();
                    if (debug) {
                        console.log(chalk_1.yellow("\n#####################################################################################\n " + (isDifference ? isDifferenceMessage : debugMessage) + "\n " + diffFilePath + "\n#####################################################################################\n"));
                    }
                    _e.label = 6;
                case 6: return [2, imageCompareOptions.returnAllCompareData ? {
                        fileName: fileName,
                        folders: __assign({ actual: actualFilePath, baseline: baselineFilePath }, (diffFilePath ? { diff: diffFilePath } : {})),
                        misMatchPercentage: misMatchPercentage,
                    } : misMatchPercentage];
            }
        });
    });
}
exports.executeImageCompare = executeImageCompare;
function makeCroppedBase64Image(base64Image, rectangles, resizeDimensions) {
    if (resizeDimensions === void 0) { resizeDimensions = constants_1.DEFAULT_RESIZE_DIMENSIONS; }
    return __awaiter(this, void 0, void 0, function () {
        var resizeValues, _a, top, right, bottom, left, height, width, x, y, canvasWidth, canvasHeight, canvas, image, ctx, sourceXStart, sourceYStart;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (typeof resizeDimensions === 'number') {
                        resizeValues = {
                            top: resizeDimensions,
                            right: resizeDimensions,
                            bottom: resizeDimensions,
                            left: resizeDimensions,
                        };
                        console.log(chalk_1.yellow("\n#####################################################################################\n WARNING:\n THE 'resizeDimensions' NEEDS TO BE AN OBJECT LIKE\n {\n    top: 10,\n    right: 20,\n    bottom: 15,\n    left: 25,\n }\n NOW IT WILL BE DEFAULTED TO\n  {\n    top: " + resizeDimensions + ",\n    right: " + resizeDimensions + ",\n    bottom: " + resizeDimensions + ",\n    left: " + resizeDimensions + ",\n }\n THIS IS DEPRACATED AND WILL BE REMOVED IN A NEW MAJOR RELEASE\n#####################################################################################\n"));
                    }
                    else {
                        resizeValues = resizeDimensions;
                    }
                    _a = __assign(__assign({}, constants_1.DEFAULT_RESIZE_DIMENSIONS), resizeValues), top = _a.top, right = _a.right, bottom = _a.bottom, left = _a.left;
                    height = rectangles.height, width = rectangles.width, x = rectangles.x, y = rectangles.y;
                    canvasWidth = width + left + right;
                    canvasHeight = height + top + bottom;
                    canvas = createCanvas(canvasWidth, canvasHeight);
                    return [4, loadImage("data:image/png;base64," + base64Image)];
                case 1:
                    image = _b.sent();
                    ctx = canvas.getContext('2d');
                    sourceXStart = x - left;
                    sourceYStart = y - top;
                    if (sourceXStart < 0) {
                        console.log(chalk_1.yellow("\n#####################################################################################\n THE RESIZE DIMENSION LEFT '" + left + "' MADE THE CROPPING GO OUT OF\n THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '" + sourceXStart + "'.\n THIS HAS BEEN DEFAULTED TO '0'\n#####################################################################################\n"));
                        sourceXStart = 0;
                    }
                    if (sourceYStart < 0) {
                        console.log(chalk_1.yellow("\n#####################################################################################\n THE RESIZE DIMENSION LEFT '" + top + "' MADE THE CROPPING GO OUT OF\n THE IMAGE BOUNDARIES RESULTING IN AN IMAGE STARTPOSITION '" + sourceYStart + "'.\n THIS HAS BEEN DEFAULTED TO '0'\n#####################################################################################\n"));
                        sourceYStart = 0;
                    }
                    ctx.drawImage(image, sourceXStart, sourceYStart, canvasWidth, canvasHeight, 0, 0, canvasWidth, canvasHeight);
                    return [2, canvas.toDataURL().replace(/^data:image\/png;base64,/, '')];
            }
        });
    });
}
exports.makeCroppedBase64Image = makeCroppedBase64Image;
function makeFullPageBase64Image(screenshotsData) {
    return __awaiter(this, void 0, void 0, function () {
        var amountOfScreenshots, canvasHeight, canvasWidth, canvas, ctx, i, _a, canvasYPosition, imageHeight, imageWidth, imageYPosition, image;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    amountOfScreenshots = screenshotsData.data.length;
                    canvasHeight = screenshotsData.fullPageHeight, canvasWidth = screenshotsData.fullPageWidth;
                    canvas = createCanvas(canvasWidth, canvasHeight);
                    ctx = canvas.getContext('2d');
                    i = 0;
                    _b.label = 1;
                case 1:
                    if (!(i < amountOfScreenshots)) return [3, 4];
                    _a = screenshotsData.data[i], canvasYPosition = _a.canvasYPosition, imageHeight = _a.imageHeight, imageWidth = _a.imageWidth, imageYPosition = _a.imageYPosition;
                    return [4, loadImage("data:image/png;base64," + screenshotsData.data[i].screenshot)];
                case 2:
                    image = _b.sent();
                    ctx.drawImage(image, 0, imageYPosition, imageWidth, imageHeight, 0, canvasYPosition, imageWidth, imageHeight);
                    _b.label = 3;
                case 3:
                    i++;
                    return [3, 1];
                case 4: return [2, canvas.toDataURL().replace(/^data:image\/png;base64,/, '')];
            }
        });
    });
}
exports.makeFullPageBase64Image = makeFullPageBase64Image;
function saveBase64Image(base64Image, filePath) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2, fs_extra_1.outputFile(filePath, base64Image, 'base64')];
        });
    });
}
exports.saveBase64Image = saveBase64Image;
function addBlockOuts(screenshot, ignoredBoxes) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, height, width, canvas, image, canvasContext;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = utils_1.getScreenshotSize(screenshot), height = _a.height, width = _a.width;
                    canvas = createCanvas(width, height);
                    return [4, loadImage("data:image/png;base64," + screenshot)];
                case 1:
                    image = _b.sent();
                    canvasContext = canvas.getContext('2d');
                    canvasContext.drawImage(image, 0, 0, width, height, 0, 0, width, height);
                    ignoredBoxes.forEach(function (ignoredBox) {
                        var ignoredBoxWidth = ignoredBox.right, ignoredBoxHeight = ignoredBox.bottom, x = ignoredBox.left, y = ignoredBox.top;
                        var ignoreCanvas = createCanvas(ignoredBoxWidth - x, ignoredBoxHeight - y);
                        var ignoreContext = ignoreCanvas.getContext('2d');
                        ignoreContext.globalAlpha = 0.5;
                        ignoreContext.fillStyle = '#39aa56';
                        ignoreContext.fillRect(0, 0, ignoredBoxWidth - x, ignoredBoxHeight - y);
                        canvasContext.drawImage(ignoreCanvas, x, y);
                    });
                    return [2, canvas.toDataURL().replace(/^data:image\/png;base64,/, '')];
            }
        });
    });
}
exports.addBlockOuts = addBlockOuts;
