// jasmine-failures-slack-reporter/index.js
exports.WebReporter = function(options) {
	var self = this;
	var testPassed = 0;
	var testTotal = 0;

	self.url = options.url;
	self.projectName = options.projectName;
	self.environment = options.environment;
	self.slackUrl = options.slackUrl;
	self.channel = options.channel;	

	var testRun =
		{
			projectName : self.projectName,
			environment : self.environment,			
			status : "passed",
			tests : []
		};

	self.specDone = function(sp) {		
		var spec = JSON.parse(JSON.stringify(sp));
		spec._endTime = new Date();
		testTotal++;
		if (spec.status === 'failed'){
			testRun.status = "failed";
		}
		if (spec.status === 'passed'){
			testPassed++;
		}
		testRun.tests.push(spec);
	};

	self.jasmineDone = function() {
		testRun.endTime = new Date();
		var pretext = testRun.projectName + " -->>  " + testRun.status.toUpperCase();
		var text = 'Environment: ' + testRun.environment.toUpperCase() + '  -->> Tests passed: ' + testPassed + '/' + testTotal;
		var color = 'danger';
		if (testRun.status === 'failed') {

			var attachments = [{ "pretext": pretext, "text": text, "color": color }];
			var payload = { "channel": self.channel, "username": "webhookbot", "attachments": attachments }

			var request = require('sync-request');
			var res = request('POST', self.slackUrl, {
				json: payload
			});
		}
	};
};
